pkg/nfd-master/nfd-master.go

```diff
@@ -825,6 +846,80 @@ func (m *nfdMaster) nfdAPIUpdateOneNode(cli k8sclient.Interface, node *corev1.No
 	return nil
 }
 
+func (m *nfdMaster) nfdAPIUpdateAllNodeFeatureGroups() error {
+	klog.V(1).InfoS("updating All NodeFeatureGroups, will process all nodeFeatureGroupRules")
+
+	nodeFeatureGroups, err := m.nfdController.featureGroupLister.List(k8sLabels.Everything())
+	if err != nil {
+		return fmt.Errorf("failed to get nodeFeatureGroupRules resources: %w", err)
+	}
+	for _, nodeFeatureGroup := range nodeFeatureGroups {
+		m.updaterPool.nfgQueue.Add(nodeFeatureGroup.Name)
+	}
+
+	return nil
+}
+
+func (m *nfdMaster) nfdAPIUpdateNodeFeatureGroup(nodeFeatureGroupName string) error {
+	klog.V(1).InfoS("processing of nodeFeatureGroup", "nodeFeatureGroup", nodeFeatureGroupName)
+	nodeGroupValidator := make(map[string]bool)
+	var err error
+
+	kubeconfig, err := utils.GetKubeconfig(m.args.Kubeconfig)
+	if err != nil {
+		return err
+	}
+	nfdClient := nfdclientset.NewForConfigOrDie(kubeconfig)
+
+	nodeFeatures, err := m.nfdController.featureLister.List(k8sLabels.Everything())
+	if err != nil {
+		return fmt.Errorf("failed to get NodeFeatures resources: %w", err)
+	}
+
+	nodeFeatureGroup, err := nfdClient.NfdV1alpha1().NodeFeatureGroups(m.namespace).Get(context.Background(), nodeFeatureGroupName, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to get NodeFeatureGroup resource: %w", err)
+	}
+
+	// Execute rules and create matching groups
+	nodes := make([]string, 0)
+	for _, rule := range nodeFeatureGroup.Spec.Rules {
+		for _, feature := range nodeFeatures {
+			match, err := nodefeaturerule.ExecuteGroupRule(&rule, &feature.Spec.Features)
+			if err != nil {
+				return fmt.Errorf("failed to evaluate rule %q: %w", rule.Name, err)
+			}
```

Comment:
Hmm, this is different from how we handle NodeFeatureRules. We should follow the same logic here. Basically process every rule only once, against a union all NodeFeatures. I.e. merge all NodeFeatures (of this node) into one mega `Features` struct and give that as input to the rule. See how `nfdAPIUpdateOneNode()` does it. Maybe it would make sense to split out the merging logic into a common function

Reply:
I revisited this func, and this is different from the `nfdAPIUpdateOneNode` yes, but we are visiting every rule just once, the thing is that for each rule we need to see all nodes to check if they are fit for the rule and add them to the Node that will become a nodeFeatureGroup in the end.

Reply:
Well, we're visiting every rule once per each node. But the features of the node must be a union of all the NodeFeature objects of that node. Think a about a rule that checks that (in pseudocode) `featureA == "foo" && featureB == "bar"`. Now, the end result must be the same, IMO, independent whether `featureA` and `featureB` are specified in one NodeFeature object or two separate ones. You might want to write rules that combine NFD's builtin features with some 3rd party extension and that's not possible if you don't merge all features of a node into a union. Plus, importantly, it's also about consistency, treat features the same way whether we're processing NodeFeatureRules or NodeFeatureGroup (rules).

Reply:
We a re visiting all nodes for each rule*

Reply:
Now I got what you meant, my traveling brain was melted. will address